### ISSUE TITLE
Change to on_train_batch_end causes averaged values of metrics.

### ISSUE THREAD

**Issue by odinsbane** [tensorflow/tensorflow#39448]
Logging batch values for the various metrics has been a very handy tool, somewhere in the transition from 1.13 to 1.14 the callback api switched from using on_batch_end, to on_train_batch_end. Along with this switch came averaged metrics.

I am trying to report my results using tensorflow in a science journal, and I significantly have to change what I am reporting because tensorflow has changed the api on a minor version change.

Is there a way to get the actual per batch metrics? 

Right now, I have different metrics for versions 1.14 and 1.13, and it is not appropriate for a professional publication where I am aiming to demonstrate something repeatable.

**Comment by ravikyram** [comment]
@odinsbane 

Will it be possible to share colab link or simple standalone code to reproduce the issue in our environment.It helps us in localizing the issue faster.Thanks!

**Comment by odinsbane** [comment]
Here is an example script that will demonstrate the difference. On tensorflow 1.12 the result is created through the first callback method and generates metrics from each individual batch. On tensorflow 1.13.1 the second method is called and it logs an averaged value.

`from tensorflow import keras

import numpy

class CustomCallback(keras.callbacks.Callback):
    def __init__(self):
        super().__init__()
        with open("batch.txt", 'w', encoding="utf8") as f:
            f.write("#batch\tacc\n")
    def on_batch_end(self, value, params):
        #called from tf 1.12
        with open("batch.txt", 'a', encoding="utf8") as f:
            f.write("%s\t%s\n"%(value, params["acc"]))
    def on_train_batch_end(self, value, params):
        #called from tf 1.13.1
        with open("batch.txt", 'a', encoding="utf8") as f:
            f.write("%s\t%s\n"%(value, params["acc"]))

def genData():
    x = []
    y = []
    
    cnets = numpy.random.randint(0, 62, (1000, 2) )
    for cnet in cnets:
        x0 = numpy.zeros((64, 64, 1))
        x0[ cnet[0] : cnet[0] + 2, cnet[1] : cnet[1] + 2, 0 ] = 1
        y0 = x0[3:61, 3:61, :]
        x0 = x0 + numpy.random.random((64, 64, 1));
        x.append(x0)
        y.append(y0)
    
    return numpy.array(x), numpy.array(y)

inp = keras.layers.Input( ( 64, 64, 1) );

next = keras.layers.Conv2D( 32, kernel_size=(3, 3),
                 activation='relu' )(inp)
next = keras.layers.Conv2D( 32, kernel_size=(3, 3),
                 activation='relu' )(next);

next = keras.layers.Conv2D( 1, kernel_size=(3, 3),
                 activation='sigmoid' )(next);

model = keras.models.Model(inputs=[inp], outputs = [next])

model.summary()
x, y = genData()

optimizer = keras.optimizers.SGD(0.01)
loss_fn = keras.losses.MSE;

model.compile(
    loss = loss_fn, 
    optimizer=optimizer, 
    metrics = ["accuracy"],
   
    )

model.fit( x, y, batch_size=10, epochs=2, verbose=0, callbacks=[CustomCallback()])
`

I can include plots of the acc value that gets logged if necessary. Also, I don't know when this change happened. I was building tf 1.13 and at some point the callbacks switched and I had to change the names of my methods.

**Comment by ravikyram** [comment]
@odinsbane 

I have tried in colab with TF version 1.12, 1.13.1, 1.14 .Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/827995a4f209e5e4622ac708474eac6f/untitled882.ipynb).Is this the expected behavior? Thanks!

**Comment by odinsbane** [comment]
Yes, that is the expected behavior. The difference that I am talking about,
is a difference in the data being written to batch.txt. In the 1.12 version
the data is noisier and reflects the values from each bach. In version
1.13+ the data is averaged during an epoch. I'm not too familiar with
colab, so I don't know a good way to plot the output.

**Comment by odinsbane** [comment]
![image](https://user-images.githubusercontent.com/3462801/81934804-791ab980-95e7-11ea-89f8-1dc80a817055.png)

This demonstrates the difference in batch metrics. The right graph changes immediately, and doesn't matter when the second epoch starts. The left graph gradually increases, and then at the start of the new epoch, it jumps to a higher value because it has been plotting an average over the epoch.

We used the differences in batch metrics to evaluate whether the batches had some sort of bias from the sources of data.

**Comment by kkedich** [comment]
It seems that this is also happening with TF 2.2.0. But with TF 2.0.0 it seems to work as expected.
Instead of printing the loss for the current batch, the `on_train_batch_end` function is printing the average loss. 

You can reproduce the expected and wrong behaviour with [this colab](https://colab.research.google.com/drive/1WAQQxBi5llzvMRrUDw4mIj0l7VQnPBTC?usp=sharing)

**Comment by fundamental** [comment]
I can also confirm the shift over to averaged metrics is disruptive to research work being done with newer versions of tensorflow (I'm likely staying with 2.0 for the time being due to this issue). It looks like the current 2.4.x nightly (from dockerhub) averages over loss which is particularly detrimental when exploring any changes that happen within a batch (e.g. learning rate variations, looking for unshuffled clumps of data being input to the system, or any other sharp transitions in loss).

I'd welcome any suggestions to be able to simply swap over to non-averaged versions of metrics when logging information out to tensorboard.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
